### ISSUE TITLE
Variable products in different languages ​​are added to the cart separately

### ISSUE THREAD

**Issue by Milakrivoruchko** [hyyan/woo-poly-integration#361]
**Can you reproduce this issue on default Wordpress theme (eg Storefront)?**
yes

**Can you reproduce this issue when all other plugins are disabled except WooCommerce, Polylang and Hyyan WooCommerce Polylang Integration?**
yes

**What product versions and settings are you using when this issue occurs?**
PHP: 7.1
WordPress: 4.7.11
WooCommerce: 3.4.4
Polylang: [state if using Polylang PRO] 2.1.5
Hyyan WooCommerce Polylang Integration: 1.2.0
Theme: Impreza (Child Theme)
Browser: Opera, Chrome, any

**Steps to Reproduce**

Add to cart variable products in different languages (Simple products are added to the cart correctly)

**What I Expected**
Variable products of the same type were added to the cart in different languages ​as one

**Note:** Variable products in these versions are translated correctly, in the additional language, data of product variations in the main language are blocked and activated. But duplicated when added to the trash.

**What was done before:**
There is a site where woo-poly does not work correctly with the translation of variations. Variations were blocked by woo-poly, Polylang combined multilingual versions correctly.
In the cart were various variable products with different variations.
Have installed a different version of WP. The recommendations of **Fixes # 187** were implemented (also when updating). Variations are handled correctly, but variable products in different languages ​​are added to the cart several times. 

**Possible actions**, write your own handler add-to-cart-validation and add-to-cart-variable_product. There are doubts.
@Jon007, please, help mine.

**Comment by Milakrivoruchko** [comment]
Test site №2:
**variable**: https://funt.demosite.biz.ua/uk/product/кава-фунт-кави/?attribute_pa_roasting=%25d1%2584%25d1%2596%25d0%25bb%25d1%258c%25d1%2582%25d1%2580
**simple**: https://funt.demosite.biz.ua/uk/product/акссесуари-укр/
result is visible in the mini-cart

https://coffee.demosite.biz.ua/

**Comment by Jon007** [comment]
Hi, 
I don't understand what you mean by 

> Variable products of the same type were added to the cart in different languages ​as one  


what problem are you experiencing?


The plugin doesn't have a with variable products and the cart, this should normally work fine.

 - products are normally shown in the current language, so if you switch languages, the system will find the new language version of the product and display that
 - if the system cannot find a translation for the product (or variation), then, just for the cart screen, the language version originally added to the cart is displayed.

Note:

Some plugins/themes could have code which interferes with the above process

This plugin does have a special link to connect the variation translations, which depends on the variations being created in the shop base language, so if the variations were created when this plugin was not activated, or the shop base language was changed, then the system would have trouble finding the linked translation of the variation.

**Comment by Milakrivoruchko** [comment]
Thank @Jon007,

please see how it looks in the cart (as **two products**) 
https://drive.google.com/open?id=1sFtJ-5q-utKhsIoy0rucdd5SBwa7lk1G

Polylang switches products correctly, woo-poly translates the variation from the main language. 

Variable products are created with an active woo-poly. Translates variational products correctly, but adds them to the cart two products, as in the screenshot. 

**I think public function addToCart ($ ID) does not handle variable products. I can not find the reason.**

What versions of WP, woo-poly, Polylang do you consider working with variational products? **What are your recommendations for the versions?**

**Comment by Jon007** [comment]
ok I see what you mean and can reproduce as follows:

1. add a variation to the cart
2. switch language
3. add the same variation to the cart in the new language
4. the cart now has two items in with different ids, but actually both of them refer to the same item and will show as two copies of the same item

This could be fixed by an extra check when adding to cart, check if the product to be added is a variation and any translation of this variation is already in the cart then increase the cart quantity instead of adding to the cart..

However I am not sure that the current behaviour actually causes any real problem does it?
This only occurs if the user deliberately adds different language versions of the same variation to the cart, which is pretty unusual.... and the user can simply remove one of the items from the cart if they want.

**Comment by Milakrivoruchko** [comment]
Yes, as you describe.

Thanks, understood.
This problem can harm if a large online store is connected to an accounting program. For example, when languages ​​are similar and there is a language problem (Russian and Ukrainian).
That is, you need to constantly make a return of the product and the client's refusal if he switched between languages. This is an extra bookkeeping posting, which causes confusion in the posting and customer dissatisfaction.

_If I write the handler code, I'll write to the post under my question._

**Comment by Jon007** [comment]
I'm still not exactly sure what the problem is or what solution you are proposing.
Clearly if the user is allowed to check out with different products in different languages they will get different product ids, it doesn't make much difference whether that's on the same order or not.

if it's a book-keeping issue and you need all the orders/postings to refer to one language then that could be added to the book-keeping code (along the same principle as this plugin cart module does).

Some people would prefer the cart module to be turned off for the sake of other plugins also modifying the cart.  If we allowed that option, then products would either disappear from the cart when switching languages (you can only check out a product in the language you added it to the cart), or possibly always show the product in original language added to cart, regardless of checkout language.

**Comment by msbir** [comment]
Hey, I am having the same issue.

Have 2 languages on the site, any product has only 1 in stock.
when creating a product (let's say in English) and then creating the same product in a different language (let's say in Spanish), I am able to add the English product to the cart and when switching to the Spanish language, I am able to add the Spanish product to the cart. (Although there is only 1 in stock).

I am not sure if this is a bug or I just don't know how to handle it.
Either way, I expect not to be able to add the product to cart if this product has already added to the cart (no matter the language of the site).

Will be great to have suggestions to handle it.

**Comment by Jon007** [comment]
There are issues with stock, see #299

In this case the problem is not so much that the same client could add the product in different languages, the real problem is that a different client could add the stock in another language as the stock is not reserved across all languages.
So any client could buy in Spanish even though the stock is notionally 'reserved' in English. I suggest following that further on #299, as the stock handling is a real bug, and quite a separate issue from this one.

**Comment by msbir** [comment]
I looked at #299 and I don't think this is my problem. 

I just got an order on the site for the English product. The Spanish product is marked as 'out of stock' and the client is not able to add to cart/purchase this product. 

My problem is when the client adding an English item to the cart and on the same session, change the site language to Spanish, the client can add the same product (only in Spanish).
